***
## Parte IV
O objetivo desta parte do trabalho é estudar o comportamento dos otimizadores de
consulta dos SGBDs através do exame e análise dos planos de execução para consultas SQL
sobre tabelas que serão fornecidos. Será bastante utilizado o comando `EXPLAIN`, que permite
visualizar todas as etapas envolvidas no processamento de uma consulta.

- _Alice Teles Lucena_
- _Igor de Souza Lima_
- _Nathália Rodrigues Machado dos Santos_

In [8]:
# Fazendo a conexão com o banco de dados
import psycopg2

conn = psycopg2.connect( 
    host="localhost",
    port="5432",
    database="tpchdb",
    user="tpch",
    password="test123"
)

cursor = conn.cursor()

print("Conexão estabelecida com o Banco de Dados !!!")

Conexão estabelecida com o Banco de Dados !!!


----------------------------------------
### Tarefa 17 - Preparação e Verificação do Ambiente
a) Execute o script movie.sql para criar as tabelas e índices e carregar os dados necessários às próximas atividades

<span style="color:orange">__Resposta Letra A:__</span> 

**OBS:** Caso você esteja executando essa célula logo depois de ter executado a mesma no notebook da parte 3, não execute novamente; o movie.sql já foi executado na parte 3, se executado novamente, a célula não terminará de carregar.

In [11]:
!psql postgresql://tpch:test123@localhost:5432/tpchdb -f movie.sql

^C
Cancel request sent
psql:movie.sql:1: ERROR:  canceling statement due to user request


----------------------------------------
### Tarefa 17 - Preparação e Verificação do Ambiente
b) Verifique no catálogo do banco de dados os seguintes metadados sobre os índices associados às tabelas e apresente-os no relatório: Nome do índice, nome da tabela, altura, número máximo de chaves por bloco, número médio de chaves por bloco, número de blocos folha, número de médio de blocos folha por chave, número médio de blocos de dados por chave, número de linhas e número de chaves distintas.

<span style="color:orange">__Resposta Letra B:__</span>

In [5]:
cursor.execute("""
    SELECT indexname
    FROM pg_indexes
    WHERE pg_indexes.tablename = 'movie';
""")
result = cursor.fetchall()
print("Nomes dos Índices: ", result)

Nomes dos Índices:  [('movie_key',), ('movie_title',), ('movie_votes',), ('movie_yr',), ('movie_director',)]


In [23]:
cursor.execute("""
    SELECT table_name
    FROM information_schema.tables
    WHERE table_schema = 'public';
""")
result = cursor.fetchall()
print("Nomes da Tabela: ", result[-3])  # result[-3] pois -> -3 = movie / -2 = actors / -2 = casting
                                        # caso o notebook da parte 3 nao tenha sido executado, que é onde rodamos os arquivos .sql dessas 3 tabelas, coloque result [-1]

Nomes da Tabela:  ('movie',)


In [14]:
cursor.execute("""
SELECT pg_size_pretty(pg_relation_size('movie'::regclass)) AS index_size
FROM pg_indexes
WHERE tablename = 'movie' AND indexname = 'movie_key';
""")
result = cursor.fetchall()
print("Altura de 'movie_key': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT pg_size_pretty(pg_relation_size('movie'::regclass)) AS index_size
FROM pg_indexes
WHERE tablename = 'movie' AND indexname = 'movie_title';
""")
result = cursor.fetchall()
print("Altura de 'movie_title': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT pg_size_pretty(pg_relation_size('movie'::regclass)) AS index_size
FROM pg_indexes
WHERE tablename = 'movie' AND indexname = 'movie_votes';
""")
result = cursor.fetchall()
print("Altura de 'movie_votes': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT pg_size_pretty(pg_relation_size('movie'::regclass)) AS index_size
FROM pg_indexes
WHERE tablename = 'movie' AND indexname = 'movie_yr';
""")
result = cursor.fetchall()
print("Altura de 'movie_yr': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT pg_size_pretty(pg_relation_size('movie'::regclass)) AS index_size
FROM pg_indexes
WHERE tablename = 'movie' AND indexname = 'movie_director';
""")
result = cursor.fetchall()
print("Altura de 'movie_director': ", result)

Altura de 'movie_key':  [('120 kB',)]
Altura de 'movie_title':  [('120 kB',)]
Altura de 'movie_votes':  [('120 kB',)]
Altura de 'movie_yr':  [('120 kB',)]
Altura de 'movie_director':  [('120 kB',)]


In [15]:
cursor.execute("""
SELECT current_setting('block_size')::numeric / (sum(attlen) + 4 + 4) AS max_chaves_por_bloco
FROM pg_index
JOIN pg_class ON pg_class.oid = pg_index.indexrelid
JOIN pg_attribute ON pg_attribute.attrelid = pg_class.oid
WHERE pg_class.relname = 'movie_key';
""")
result = cursor.fetchall()
print("Número Máximo de Chaves por Bloco em 'movie_key': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT current_setting('block_size')::numeric / (sum(attlen) + 4 + 4) AS max_chaves_por_bloco
FROM pg_index
JOIN pg_class ON pg_class.oid = pg_index.indexrelid
JOIN pg_attribute ON pg_attribute.attrelid = pg_class.oid
WHERE pg_class.relname = 'movie_title';
""")
result = cursor.fetchall()
print("Número Máximo de Chaves por Bloco em 'movie_title': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT current_setting('block_size')::numeric / (sum(attlen) + 4 + 4) AS max_chaves_por_bloco
FROM pg_index
JOIN pg_class ON pg_class.oid = pg_index.indexrelid
JOIN pg_attribute ON pg_attribute.attrelid = pg_class.oid
WHERE pg_class.relname = 'movie_votes';
""")
result = cursor.fetchall()
print("Número Máximo de Chaves por Bloco em 'movie_votes': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT current_setting('block_size')::numeric / (sum(attlen) + 4 + 4) AS max_chaves_por_bloco
FROM pg_index
JOIN pg_class ON pg_class.oid = pg_index.indexrelid
JOIN pg_attribute ON pg_attribute.attrelid = pg_class.oid
WHERE pg_class.relname = 'movie_yr';
""")
result = cursor.fetchall()
print("Número Máximo de Chaves por Bloco em 'movie_yr': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT current_setting('block_size')::numeric / (sum(attlen) + 4 + 4) AS max_chaves_por_bloco
FROM pg_index
JOIN pg_class ON pg_class.oid = pg_index.indexrelid
JOIN pg_attribute ON pg_attribute.attrelid = pg_class.oid
WHERE pg_class.relname = 'movie_director';
""")
result = cursor.fetchall()
print("Número Máximo de Chaves por Bloco em 'movie_director': ", result)

Número Máximo de Chaves por Bloco em 'movie_key':  [(Decimal('819.2000000000000000'),)]
Número Máximo de Chaves por Bloco em 'movie_title':  [(Decimal('1170.2857142857142857'),)]
Número Máximo de Chaves por Bloco em 'movie_votes':  [(Decimal('682.6666666666666667'),)]
Número Máximo de Chaves por Bloco em 'movie_yr':  [(Decimal('819.2000000000000000'),)]
Número Máximo de Chaves por Bloco em 'movie_director':  [(Decimal('819.2000000000000000'),)]


In [16]:
cursor.execute("""
SELECT (pg_relation_size('movie') / pg_stat_get_live_tuples('movie'::regclass))::numeric AS avg_chaves_por_bloco
FROM pg_class
WHERE relname = 'movie';
""")
result = cursor.fetchall()
print("Número Médio de Chaves por Bloco: ", result)

Número Médio de Chaves por Bloco:  [(Decimal('66'),)]


In [17]:
cursor.execute("""
SELECT pg_relation_size('movie_key') / current_setting('block_size')::numeric AS numero_de_blocos_folha
FROM pg_class
WHERE relname = 'movie_key';
""")
result = cursor.fetchall()
print("Número de Blocos Folha em 'movie_key': ", result)
# ---------------------------------------------
cursor.execute("""
SELECT pg_relation_size('movie_title') / current_setting('block_size')::numeric AS numero_de_blocos_folha
FROM pg_class
WHERE relname = 'movie_title';
""")
result = cursor.fetchall()
print("Número de Blocos Folha em 'movie_title': ", result)
# ---------------------------------------------
cursor.execute("""
SELECT pg_relation_size('movie_votes') / current_setting('block_size')::numeric AS numero_de_blocos_folha
FROM pg_class
WHERE relname = 'movie_votes';
""")
result = cursor.fetchall()
print("Número de Blocos Folha em 'movie_votes': ", result)
# ---------------------------------------------
cursor.execute("""
SELECT pg_relation_size('movie_yr') / current_setting('block_size')::numeric AS numero_de_blocos_folha
FROM pg_class
WHERE relname = 'movie_yr';
""")
result = cursor.fetchall()
print("Número de Blocos Folha em 'movie_yr': ", result)
# ---------------------------------------------
cursor.execute("""
SELECT pg_relation_size('movie_director') / current_setting('block_size')::numeric AS numero_de_blocos_folha
FROM pg_class
WHERE relname = 'movie_director';
""")
result = cursor.fetchall()
print("Número de Blocos Folha em 'movie_director': ", result)

Número de Blocos Folha em 'movie_key':  [(Decimal('7.0000000000000000'),)]
Número de Blocos Folha em 'movie_title':  [(Decimal('12.0000000000000000'),)]
Número de Blocos Folha em 'movie_votes':  [(Decimal('11.0000000000000000'),)]
Número de Blocos Folha em 'movie_yr':  [(Decimal('5.0000000000000000'),)]
Número de Blocos Folha em 'movie_director':  [(Decimal('6.0000000000000000'),)]


In [18]:
cursor.execute("""
SELECT (reltuples / (relpages * 0.8)) AS media_blocos_folha_por_chave
FROM pg_class
WHERE relname = 'movie_key';
""")
result = cursor.fetchall()
print("Número Médio de Blocos Folha por Chave em 'movie_key': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT (reltuples / (relpages * 0.8)) AS media_blocos_folha_por_chave
FROM pg_class
WHERE relname = 'movie_title';
""")
result = cursor.fetchall()
print("Número Médio de Blocos Folha por Chave em 'movie_title': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT (reltuples / (relpages * 0.8)) AS media_blocos_folha_por_chave
FROM pg_class
WHERE relname = 'movie_votes';
""")
result = cursor.fetchall()
print("Número Médio de Blocos Folha por Chave em 'movie_votes': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT (reltuples / (relpages * 0.8)) AS media_blocos_folha_por_chave
FROM pg_class
WHERE relname = 'movie_yr';
""")
result = cursor.fetchall()
print("Número Médio de Blocos Folha por Chave em 'movie_yr': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT (reltuples / (relpages * 0.8)) AS media_blocos_folha_por_chave
FROM pg_class
WHERE relname = 'movie_director';
""")
result = cursor.fetchall()
print("Número Médio de Blocos Folha por Chave em 'movie_director': ", result)

Número Médio de Blocos Folha por Chave em 'movie_key':  [(329.28571428571433,)]
Número Médio de Blocos Folha por Chave em 'movie_title':  [(192.08333333333334,)]
Número Médio de Blocos Folha por Chave em 'movie_votes':  [(209.54545454545453,)]
Número Médio de Blocos Folha por Chave em 'movie_yr':  [(461.0,)]
Número Médio de Blocos Folha por Chave em 'movie_director':  [(384.1666666666667,)]


In [19]:
cursor.execute("""
SELECT (pg_relation_size('movie') / pg_stat_get_live_tuples('movie'::regclass))::numeric AS avg_blocos_dados_por_chave
FROM pg_class
WHERE relname = 'movie';
""")
result = cursor.fetchall()
print("Número Médio de Blocos de Dados por Chave: ", result)

Número Médio de Blocos de Dados por Chave:  [(Decimal('66'),)]


In [20]:
cursor.execute("""
SELECT reltuples AS numero_linhas
FROM pg_class
WHERE relname = 'movie_key' AND relkind = 'i';
""")
result = cursor.fetchall()
print("Número de Linhas no Índice 'movie_key': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT reltuples AS numero_linhas
FROM pg_class
WHERE relname = 'movie_title' AND relkind = 'i';
""")
result = cursor.fetchall()
print("Número de Linhas no Índice 'movie_title': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT reltuples AS numero_linhas
FROM pg_class
WHERE relname = 'movie_votes' AND relkind = 'i';
""")
result = cursor.fetchall()
print("Número de Linhas no Índice 'movie_votes': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT reltuples AS numero_linhas
FROM pg_class
WHERE relname = 'movie_yr' AND relkind = 'i';
""")
result = cursor.fetchall()
print("Número de Linhas no Índice 'movie_yr': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT reltuples AS numero_linhas
FROM pg_class
WHERE relname = 'movie_director' AND relkind = 'i';
""")
result = cursor.fetchall()
print("Número de Linhas no Índice 'movie_director': ", result)
# --------------------------------------------------------------------

Número de Linhas no Índice 'movie_key':  [(1844.0,)]
Número de Linhas no Índice 'movie_title':  [(1844.0,)]
Número de Linhas no Índice 'movie_votes':  [(1844.0,)]
Número de Linhas no Índice 'movie_yr':  [(1844.0,)]
Número de Linhas no Índice 'movie_director':  [(1844.0,)]


In [21]:
cursor.execute("""
SELECT COUNT(DISTINCT id) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `id` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT title) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `title` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT yr) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `yr` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT score) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `score` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT votes) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `votes` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT director) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `director` distintas: ", result)
# --------------------------------------------------------------------

Número de Chaves `id` distintas:  [(1844,)]
Número de Chaves `title` distintas:  [(1833,)]
Número de Chaves `yr` distintas:  [(75,)]
Número de Chaves `score` distintas:  [(66,)]
Número de Chaves `votes` distintas:  [(1481,)]
Número de Chaves `director` distintas:  [(820,)]


----------------------------------------
### Tarefa 18 - Consultas por intervalo e índices secundários
a) Escreva uma consulta em SQL sobre o atributo VOTES da tabela MOVIE que recupera um número pequeno de tuplas (<10 tuplas); Execute o comando `EXPLAIN` sobre esta consulta e apresente os resultados.

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra A:__</span>

In [24]:
cursor.execute("EXPLAIN (ANALYZE) SELECT * FROM movie WHERE votes = 10 LIMIT 9")

result = cursor.fetchall()

for row in result:
        print( row)

('Limit  (cost=0.28..8.29 rows=1 width=30) (actual time=0.017..0.018 rows=0 loops=1)',)
('  ->  Index Scan using movie_votes on movie  (cost=0.28..8.29 rows=1 width=30) (actual time=0.016..0.016 rows=0 loops=1)',)
('        Index Cond: (votes = 10)',)
('Planning Time: 0.194 ms',)
('Execution Time: 0.050 ms',)


b) Escreva uma consulta em SQL sobre o atributo VOTES da tabela MOVIE que recupera um número grande de tuplas (>80% das tuplas). Execute o comando `EXPLAIN` sobre esta consulta e apresente os resultados.

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra B:__</span>

In [25]:
cursor.execute("EXPLAIN (ANALYZE) SELECT * FROM MOVIE LIMIT (SELECT COUNT(*) * 0.8 FROM MOVIE)::integer;")

result = cursor.fetchall()

for row in result:
    print(row)

('Limit  (cost=38.06..41.40 rows=184 width=30) (actual time=1.358..2.470 rows=1475 loops=1)',)
('  InitPlan 1 (returns $0)',)
('    ->  Aggregate  (cost=38.05..38.06 rows=1 width=32) (actual time=1.316..1.319 rows=1 loops=1)',)
('          ->  Seq Scan on movie movie_1  (cost=0.00..33.44 rows=1844 width=0) (actual time=0.031..0.820 rows=1844 loops=1)',)
('  ->  Seq Scan on movie  (cost=0.00..33.44 rows=1844 width=30) (actual time=0.027..0.662 rows=1475 loops=1)',)
('Planning Time: 0.438 ms',)
('Execution Time: 2.815 ms',)


c) Explique porque o índice sobre VOTES não é sempre usado nas consultas sobre este atributo.

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra C:__</span>

A escolha de usar ou não um índice depende de muitos fatores, como tamanho da tabela, distribuição dos dados, cardinalidade da coluna, estatísticas atualizadas, entre outros. O otimizador de consultas do banco de dados decide qual plano de execução é o mais eficiente com base nessas informações. Nesse caso, a consulta precisa processar todas as linhas da tabela para calcular o valor da subconsulta e determinar o limite de linhas. O que torna mais viável usar a leitura sequencial no segundo código, diferente do que é feito na primeira.

----------------------------------------
### Tarefa 19 - Comparações de operadores de agregação.
Considere as seguintes consultas em SQL, sobre o atributo VOTES, as quais são equivalentes:
> `SELECT title FROM movie WHERE votes >= (SELECT MAX(votes) FROM movie);`
>
> `SELECT title FROM movie WHERE votes >= ALL (SELECT votes FROM movie) ;`

a) Apresente o resultado do comando explain sobre as duas consultas acima

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra A:__</span>

In [26]:
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE votes >= (SELECT MAX(votes) FROM movie);
""")
result = cursor.fetchall()
print(result, "\n\n")
# ----------------------------------------------
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE votes >= ALL (SELECT votes FROM movie) ;
""")
result = cursor.fetchall()
print(result)

[('Index Scan using movie_votes on movie  (cost=0.62..35.38 rows=615 width=16)',), ('  Index Cond: (votes >= $1)',), ('  InitPlan 2 (returns $1)',), ('    ->  Result  (cost=0.33..0.34 rows=1 width=4)',), ('          InitPlan 1 (returns $0)',), ('            ->  Limit  (cost=0.28..0.33 rows=1 width=4)',), ('                  ->  Index Only Scan Backward using movie_votes on movie movie_1  (cost=0.28..94.55 rows=1844 width=4)',), ('                        Index Cond: (votes IS NOT NULL)',)] 


[('Seq Scan on movie  (cost=0.00..43620.99 rows=922 width=16)',), ('  Filter: (SubPlan 1)',), ('  SubPlan 1',), ('    ->  Materialize  (cost=0.00..42.66 rows=1844 width=4)',), ('          ->  Seq Scan on movie movie_1  (cost=0.00..33.44 rows=1844 width=4)',)]


b) Existe alguma diferença entre os planos de consultas? Qual das duas é mais eficiente? Explique.

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra B:__</span>

A consulta 1 utiliza o Index Scan usando o index 'movie_votes' enquanto a 2 utiliza o Seq Scan, ambos na tabela 'movie'. No plano de consulta da primeira consulta, podemos ver a linha "Index Scan using movie_votes", o que indica que está sendo usado um índice chamado "movie_votes". Além disso, a linha "Index Cond: (votes >= $1)" mostra que está sendo feita uma verificação de condição no índice. Já no plano de consulta da segunda consulta, podemos ver a linha "Seq Scan on movie", o que indica uma varredura sequencial na tabela "movie". Além disso, a linha "Filter: (SubPlan 1)" mostra que há uma filtragem sendo aplicada.

A primeira consulta é mais eficiente pois utiliza um índice para realizar a verificação da condição, o que geralmente é mais rápido do que uma varredura sequencial na tabela inteira, mantendo seu custo entre 0.62 e 35.38; A segunda consulta precisa percorrer todas as linhas da tabela para verificar a condição, o que se torna mais lento, mantendo seu custo entre 0.00 e 43620.99.

----------------------------------------
### Tarefa 20 - Consultas com Junção e Seleção
Considere as duas consultas equivalentes em SQL a seguir, as quais retornam os filmes com
mais votos que “Star Wars”
> `SELECT title FROM movie WHERE votes > (SELECT votes FROM movie WHERE title = 'Star Wars');`
>
> `SELECT m1.title FROM movie m1, movie m2 WHERE m1.votes > m2.votes AND m2.title = 'Star Wars';`

a) Apresente o resultado do comando explain sobre as duas consultas acima

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra A:__</span>

In [27]:
cursor.execute("EXPLAIN SELECT title FROM movie WHERE votes > (SELECT votes FROM movie WHERE title = 'Star Wars');")
result = cursor.fetchall()
print(result, "\n\n")
# ----------------------------------------------
cursor.execute("EXPLAIN SELECT m1.title FROM movie m1, movie m2 WHERE m1.votes > m2.votes AND m2.title = 'Star Wars';")
result = cursor.fetchall()
print(result)

[('Index Scan using movie_votes on movie  (cost=8.57..43.34 rows=615 width=16)',), ('  Index Cond: (votes > $0)',), ('  InitPlan 1 (returns $0)',), ('    ->  Index Scan using movie_title on movie movie_1  (cost=0.28..8.29 rows=1 width=4)',), ("          Index Cond: ((title)::text = 'Star Wars'::text)",)] 


[('Nested Loop  (cost=0.56..49.49 rows=615 width=16)',), ('  ->  Index Scan using movie_title on movie m2  (cost=0.28..8.29 rows=1 width=4)',), ("        Index Cond: ((title)::text = 'Star Wars'::text)",), ('  ->  Index Scan using movie_votes on movie m1  (cost=0.28..35.04 rows=615 width=20)',), ('        Index Cond: (votes > m2.votes)',)]


b) Existe alguma diferença entre os planos de consultas? Qual das duas é mais eficiente? Explique.

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra B:__</span>

Na primeira consulta, há um índice sendo usado para a coluna "votes" na tabela "movie" usando o operador ">" na condição "Index Cond: (votes > $0)". Em seguida, há um plano de inicialização (InitPlan) para buscar o número de votos do filme "Star Wars" usando um índice na coluna "title" da tabela "movie" na subconsulta "Index Scan using movie_title on movie movie_1". Essa subconsulta é executada uma vez para obter o número de votos de "Star Wars" e depois usado na condição da consulta principal.

Na segunda consulta, há um "Nested Loop" que combina as tabelas "movie m1" e "movie m2". Primeiro, é realizado um índice para obter o filme com título "Star Wars" na tabela "movie m2" usando o índice na coluna "title" na condição "Index Cond: ((title)::text = 'Star Wars'::text)". Em seguida, há um índice sendo usado para a coluna "votes" na tabela "movie m1" usando a condição "Index Cond: (votes > m2.votes)", onde "m2.votes" é o número de votos de "Star Wars" obtido na primeira etapa.

A segunda consulta é mais eficiente, pois ela usa um Nested Loop para combinar as tabelas diretamente, evitando a necessidade de uma subconsulta separada para obter o número de votos de "Star Wars". Além disso, a segunda consulta pode fazer uso do índice na coluna "votes" diretamente na condição de junção, enquanto a primeira consulta precisa usar um índice para obter o número de votos de "Star Wars" antes de realizar a comparação.

----------------------------------------
### Tarefa 21 - Casamento de Strings e Índices
Considere as seguintes consultas SQL sobre o atributo TITLE usando o operador LIKE.
> `SELECT title FROM movie WHERE title LIKE 'I%';`
>
> `SELECT title FROM movie WHERE substr(title, 1, 1) = 'I';`
>
> `SELECT title FROM movie WHERE title LIKE '%A';`

a) Apresente o resultado do comando explain sobre as três consultas acima

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra A:__</span>

In [28]:
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE title LIKE 'I%';
""")
result = cursor.fetchall()
print(result, "\n\n")
# ----------------------------------------------
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE substr(title, 1, 1) = 'I';
""")
result = cursor.fetchall()
print(result, "\n\n")
# ----------------------------------------------
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE title LIKE '%A';
""")
result = cursor.fetchall()
print(result)

[('Seq Scan on movie  (cost=0.00..38.05 rows=18 width=16)',), ("  Filter: ((title)::text ~~ 'I%'::text)",)] 


[('Seq Scan on movie  (cost=0.00..42.66 rows=9 width=16)',), ("  Filter: (substr((title)::text, 1, 1) = 'I'::text)",)] 


[('Seq Scan on movie  (cost=0.00..38.05 rows=37 width=16)',), ("  Filter: ((title)::text ~~ '%A'::text)",)]


b) Qual das três apresenta o menor custo? Porque?

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra B:__</span>

As consultas 1 e 3 apresentam o menor custo porque utilizam o operador LIKE com uma busca simples por padrão no início ou no final da string. Nesses casos, o otimizador do banco de dados pode usar uma busca sequencial (Seq Scan) para examinar todas as linhas da tabela e filtrar apenas as que correspondem ao padrão. Já a consulta 2 usa a função substr para extrair o primeiro caractere da coluna title e depois compara com 'I'. Embora a função substr seja eficiente, ela precisa ser aplicada em todas as linhas da tabela, o que pode ser um pouco mais custoso em comparação com o operador LIKE direto.

c) O índice sobre TITLE foi usado para todas elas? Justifique.

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra C:__</span>

O índice sobre o atributo TITLE não foi usado para nenhuma das consultas. Isso pode ser inferido pelos resultados do comando EXPLAIN, que mostram um "Seq Scan" (varredura sequencial) em vez de um "Index Scan" (varredura por índice). O otimizador de consulta determina a estratégia de acesso aos dados com base em várias considerações, incluindo a presença de índices adequados e o tamanho da tabela. No caso dessas consultas, é possível que o otimizador tenha decidido que o uso de um índice não seria eficiente o suficiente em comparação com uma varredura sequencial completa da tabela.

----------------------------------------
### Tarefa 22 - Verificação da hipótese de distribuição uniforme na estimativa de seletividade
Considere as seguintes consultas sobre o atributo TITLE da tabela MOVIE
> `SELECT title FROM movie WHERE votes < 1000;`
>
> `SELECT title FROM movie WHERE votes > 40000`

a) Apresente o resultado do comando explain sobre as duas consultas acima. Explique o resultado.

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra A:__</span>

In [30]:
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE votes < 1000;
""")
result = cursor.fetchall()
print(result, "\n\n")
# ----------------------------------------------
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE votes > 40000;
""")
result = cursor.fetchall()
print(result)

[('Index Scan using movie_votes on movie  (cost=0.28..20.02 rows=328 width=16)',), ('  Index Cond: (votes < 1000)',)] 


[('Index Scan using movie_votes on movie  (cost=0.28..8.42 rows=8 width=16)',), ('  Index Cond: (votes > 40000)',)]


**Explicando o resultado:** A primeira consulta utiliza o índice "movie_votes" através da operação "Index Scan", o custo estimado da consulta é de 0.28 a 20.02, estima-se que a consulta retorna 328 linhas e a condição do índice é "(votes < 1000)".
Já a segunda consulta, também utiliza o índice "movie_votes" através da operação "Index Scan", o custo estimado da consulta é de 0.28 a 8.42, estima-se que a consulta retorna 8 linhas e condição do índice é "(votes > 40000)".

b) Compare o número de tuplas selecionadas por cada consulta. Qual das duas tem a menor seletividade?

__Entrega:__ Relatório com as respostas das questões.

<span style="color:orange">__Resposta Letra B:__</span>

Comparando o número de tuplas selecionadas por cada consulta, vemos que a segunda consulta tem uma menor seletividade. Isso é indicado pelo fato de que a estimativa de linhas retornadas é menor para essa consulta (8 linhas) em comparação com a primeira consulta (328 linhas).